# Bedingte Entropien von multivariaten Zufallsprozessen

## Bedingte Wahrscheinlichkeiten 1

Wir betrachten einen bivariaten Prozess. Die vier Jahreszeiten $\mathbb{J} = \{ \text{Frühling}, \text{Sommer}, \text{Herbst}, \text{Winter} \}$ treten jeweils mit der Wahrscheinlichkeit $\Pr \{ J_{\mu} = 0.25 \}$ auf. Dazu betrachten wir die Temperaturen $\mathbb{T} = \{ -10°C, +10°C, +20°C \}$. Die bedingten Wahrscheinlichkeiten lauten


| $\Pr\{ {\cal T}|{\cal J} \}$ | Frühling | Sommer | Herbst | Winter |
| ----- | ----- | ----- | ----- | ----- | 
| -10°C |  0.1  |  0.0  |  0.1  |  0.8  |
| +10°C |  0.6  |  0.1  |  0.6  |  0.2  |
| +20°C |  0.3  |  0.9  |  0.3  |   0   | 


Die bedingte Entropie $H({\cal T} \mid {\cal J})$ ist definiert als

\begin{align*}
H({\cal T} \mid {\cal J})
&= \mathrm{E} \big\{ \log_2 \Pr \{ {\cal T}=t \mid {\cal J}=j \} \big\} \\
&= - \sum_{j \in \mathbb{J}} \sum_{t \in \mathbb{T}} \Pr \{ {\cal T}=t, {\cal J}=j \} \cdot \log_2 \Pr \{ {\cal T}=t \mid {\cal J}=j \} \\
&= - \sum_{j \in \mathbb{J}} \Pr\{ {\cal J} = j \} \cdot \sum_{t \in \mathbb{T}} \Pr \{ {\cal T}=t \mid {\cal J}=j \} \cdot \log_2 \Pr \{ {\cal T}=t \mid {\cal J}=j \} \\
&= - \frac{1}{4} \cdot  \sum_{j \in \mathbb{J}} \sum_{t \in \mathbb{T}} \Pr \{ {\cal T}=t \mid {\cal J}=j \} \cdot \log_2 \Pr \{ {\cal T}=t \mid {\cal J}=j \}
\end{align*}

Für $H({\cal T} \mid {\cal J})=0$ wäre die Temperatur vollständig durch die Jahreszeit bestimmt, d.h. bei bekannter Jahreszeit gibt es keine Unsicherheit hinsichtlich der Temperatur. Für $H({\cal T} \mid {\cal J})=H({\cal T})$ ist die Unsicherheit hinsichtlich der Temperatur maximal, d.h. die Jahreszeit enthält keine Information über die Temperatur, beide Größen wären statistisch unabhängig voneinander. 

In [1]:
import numpy as np

# Definition der bedingten Wahrscheinlichkeiten Temperatur gegeben Jahreszeit 
Pr_T_J = np.array([[0.1, 0, 0.1, 0.8],[0.6, 0.1, 0.6, 0.2],[0.3, 0.9, 0.3, 0.0]])

# Berechnung der bedingten Entropie H(T | J)  (Nullen werden nicht berücksichtigt da 0 * log(0) = 0)
H_T_J = 0.25 * Pr_T_J * np.log2(Pr_T_J, out=np.zeros_like(Pr_T_J), where=(Pr_T_J!=0))
H_T_J = - np.sum(H_T_J)

print("Die bedingte Entropie hat den Wert H(T | J) = %g bit.\n" % (H_T_J))

Die bedingte Entropie hat den Wert H(T | J) = 0.945462 bit.



## Verbundwahrscheinlichkeiten 

Die Verbundwahrscheinlichkeiten können über die Beziehung 

\begin{equation}
    \Pr\{ {\cal T}=t, {\cal J}=j\} = \Pr\{ {\cal T}=t \mid {\cal J}=j\} \cdot \Pr\{ {\cal J}=j\}
\end{equation}

berechnet werden. Die zugehörige Verbundentropie ist zu 

\begin{align*}
H({\cal T}, {\cal J}) 
&= \mathrm{E} \big\{ \log_2 \Pr\{ {\cal T}=t, {\cal J}=j\} \big\} \\
&= - \sum_{j \in \mathbb{J}} \sum_{t \in \mathbb{T}} \Pr \{ {\cal T}=t, {\cal J}=j \} \cdot \log_2 \Pr \{ {\cal T}=t, {\cal J}=j \} 
\end{align*}

definiert.

In [2]:
# Berechnung der Verbundwahrscheinlichkeiten
Pr_TJ = Pr_T_J * 0.25

print("Die Verbundwahrscheinlichkeiten lauten \n", Pr_TJ, "\n")

Die Verbundwahrscheinlichkeiten lauten 
 [[0.025 0.    0.025 0.2  ]
 [0.15  0.025 0.15  0.05 ]
 [0.075 0.225 0.075 0.   ]] 



In [3]:
# Berechnung der Verbundentropie (Nullen werden nicht berücksichtigt da 0 * log(0) = 0)
H_TJ = Pr_TJ * np.log2(Pr_TJ, out=np.zeros_like(Pr_TJ), where=(Pr_TJ!=0))
H_TJ = - np.sum(H_TJ)

print("Die Verbundentropie beträgt %g bit.\n" % (H_TJ))

Die Verbundentropie beträgt 2.94546 bit.



## Marginale Wahrscheinlichkeiten 

Die marginalen Wahrscheinlichkeiten können über die einfchen Beziehungen

\begin{align}
    \Pr\{ {\cal T}=t \} &= \sum_{ j \in \mathbb{J}} \Pr\{ {\cal T}=t, {\cal J}=j\} \\
    \Pr\{ {\cal J}=j \} &= \sum_{ t \in \mathbb{T}} \Pr\{ {\cal T}=t, {\cal J}=j\} 
\end{align}

berechnet werden. Die Wahrscheinlichkeiten der Temperaturen werden demnach als Summe über die Spalten (je Zeile) berechnet, die der Jahrenzeiten als Summe über die Zeilen (je Spalte).

In [4]:
# marginale Wahrscheinlichkeiten für die Temperaturen
Pr_T = np.sum(Pr_TJ,axis=1)
Pr_J = np.sum(Pr_TJ,axis=0)

# Entropien für Temperatur und Jahreszeit
H_T = - Pr_T @ np.transpose(np.log2(Pr_T))
H_J = - Pr_J @ np.transpose(np.log2(Pr_J))

print("Die Wahrscheinlichkeiten der Temperaturen lauten \n", Pr_T)
print("Die Entropie der Temperatur beträgt %g bit. \n" %(H_T))

print("Zur Kontrolle: Die Wahrscheinlichkeiten der Jahrenzeiten sind \n", Pr_J)
print("Die Entropie der Jahreszeiten beträgt %g bit. \n" %(H_J))

Die Wahrscheinlichkeiten der Temperaturen lauten 
 [0.25  0.375 0.375]
Die Entropie der Temperatur beträgt 1.56128 bit. 

Zur Kontrolle: Die Wahrscheinlichkeiten der Jahrenzeiten sind 
 [0.25 0.25 0.25 0.25]
Die Entropie der Jahreszeiten beträgt 2 bit. 



## Bedingte Wahrscheinlichkeiten 2

Basierend auf den marginalen und den Verbundwahrscheinlichkeiten lassen sich nun auch die bedingten Wahrscheinlichkeiten

\begin{equation}
    \Pr\{ {\cal J} \mid {\cal T} \} = \frac{\Pr\{ {\cal T},  {\cal J} \}} {\Pr\{ {\cal T} \}}
\end{equation}

bestimmen. Sie geben an, mit welcher Wahrscheinlichkeit eine Jahreszeit bei gegebener Temperatur vorkommt. Die bedingte Entropie $H({\cal J} \mid {\cal T})$ ergibt sich zu

\begin{align*}
    H({\cal J} \mid {\cal T}) 
    &= \mathrm{E} \big\{ \log_2 \Pr\{ {\cal J} \mid {\cal T} \} \big\} \\
    &= - \sum_{j \in \mathbb{J}} \sum_{t \in \mathbb{T}} \Pr \{ {\cal T}=t, {\cal J}=j \} \cdot \log_2 \Pr \{ {\cal J}=j \mid {\cal T}=t \} \\
    &= - \sum_{t \in \mathbb{T}} \Pr\{ {\cal T} = t \} \cdot \sum_{j \in \mathbb{J}} \Pr \{ {\cal J}=j \mid {\cal T}=t \} \cdot \log_2 \Pr \{ {\cal J}=j \mid {\cal T}=t \} \\
\end{align*}

In [5]:
# Bedingte Wahrscheinlichkeit berechnen
Pr_J_T = Pr_TJ / np.tile(np.reshape(Pr_T,(3, 1)),(1,4))

# Berechnung der bedingten Entropie H( J | T)  (Nullen werden nicht berücksichtigt da 0 * log(0) = 0)
H_J_T = Pr_TJ * np.log2(Pr_J_T, out=np.zeros_like(Pr_J_T), where=(Pr_J_T!=0))
H_J_T = - np.sum(H_J_T)

print("Die bedingten Wahrscheinlichkeiten Pr{J | T} lauten \n", Pr_J_T, "\n")

print("Die bedingte Entropie H(J | T) hat den Wert %g bit.\n" %(H_J_T))

Die bedingten Wahrscheinlichkeiten Pr{J | T} lauten 
 [[0.1        0.         0.1        0.8       ]
 [0.4        0.06666667 0.4        0.13333333]
 [0.2        0.6        0.2        0.        ]] 

Die bedingte Entropie H(J | T) hat den Wert 1.38418 bit.



Der Vergleich der bedingten Entropien mit den nicht bedingten Entropien verdeutlicht, dass Seiteninformation die Unsicherheit verringert. So nimmt die Entropie von 1.56 bit auf 0.945 bit ab, wenn die Jahreszeit bekannt ist. Für die Jahreszeiten nimmt die Entropie von 2 bit auf 1.38 bit bei bekannter Temperatur ab. Hieraus können wir für das betrachtete fiktive Beispiel schließen, dass die Temperatur weniger über die Jahreszeit auszusagen scheint als umgekehrt.

## Transinformation 

Die Transinformation, auch als wechselseitige Information (mutual information) bezeichnet, gibt die gemeinsame Information von Jahreszeiten und Temperatur an. Sie ist zu

\begin{align}
    I({\cal T};{\cal J})
    &= H({\cal T}) + H({\cal J}) - H({\cal T},{\cal J})
     = H({\cal T}) - H({\cal T} \mid {\cal J})
     = H({\cal J}) - H({\cal J} \mid {\cal T})
\end{align}

definiert. Je größer die Transinformation, desto ähnlicher sind sich die betrachteten Prozesse. Dabei ist die Transinformation durch das Minimum von $H({\cal J})$ und $H({\cal T})$ nach oben begrenzt.

In [6]:
I_JT = 2 - H_J_T

print("Die Transinformation hat den Wert I(J;T)=%g bit.\n" %(I_JT))

Die Transinformation hat den Wert I(J;T)=0.615816 bit.

